File: packages/kotlin-markdown-core/src/commonMain/kotlin/com/nouprax/markdown/core/model/Code.kt
package com.nouprax.markdown.core

/** An inline code span. */
public class Code internal constructor(
    override val id: MarkupID,
    override val revision: ULong,
    override val scope: Scope,
    /** Always [PlacementMode.EMBEDDED]: a code span is inline by definition.
     *
     * The field is here because placement is one question asked of the code,
     * formula, and directive kinds alike, and [Formula] is the only one whose
     * answer varies. */
    public val mode: PlacementMode,
    /** The span's text after normalization.
     *
     * Each line ending inside it becomes one space, and one space is dropped
     * from each end when the content has both and is not entirely spaces. It
     * is therefore not always the source between the backticks. */
    public val literal: String,
) : Markup {
    override fun <Result> accept(visitor: MarkupVisitor<Result>): Result = visitor.visit(this)

    override fun equals(other: Any?): Boolean = markupEquals(this, other)

    override fun hashCode(): Int = markupHashCode(this)
}

File: packages/es-markdown-core/src/model/code.ts
import type { PlacementMode } from "../values.js";
import type { MarkupBase } from "./base.js";

/** An inline code span. */
export interface Code extends MarkupBase<"code"> {
    /** Always `embedded`.
     *
     * The standalone form of code is {@link CodeBlock}, a different kind. */
    readonly mode: PlacementMode;
    /** The text between the backticks, normalized as CommonMark requires:
     * every line ending becomes one space, and one leading and one trailing
     * space are dropped together when the span holds anything other than
     * spaces.
     *
     * Nothing in it is decoded — an entity or a backslash inside a span is
     * literal text. */
    readonly literal: string;
}
